File: app/src/main/java/com/kickstarter/mock/factories/ConfigFactory.java
package com.kickstarter.mock.factories;

import com.kickstarter.libs.Config;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;

import androidx.annotation.NonNull;

public final class ConfigFactory {
  private ConfigFactory() {}

  public static @NonNull Config config() {
    final Config.LaunchedCountry US = Config.LaunchedCountry.builder()
      .name("US")
      .currencyCode("USD")
      .currencySymbol("$")
      .trailingCode(true)
      .build();

    final Config.LaunchedCountry GB = Config.LaunchedCountry.builder()
      .name("GB")
      .currencyCode("GBP")
      .currencySymbol("£")
      .trailingCode(false)
      .build();

    final Config.LaunchedCountry CA = Config.LaunchedCountry.builder()
      .name("CA")
      .currencyCode("CAD")
      .currencySymbol("$")
      .trailingCode(true)
      .build();

    return Config.builder()
      .countryCode("US")
      .features(Collections.emptyMap())
      .launchedCountries(Arrays.asList(US, GB, CA))
      .build();
  }

  public static @NonNull Config configForUSUser() {
    return config();
  }

  public static @NonNull Config configForCAUser() {
    return config().toBuilder()
      .countryCode("CA")
      .build();
  }

  public static @NonNull Config configForITUser() {
    return config().toBuilder()
      .countryCode("IT")
      .build();
  }

  public static @NonNull Config configWithFeatureEnabled(final @NonNull String featureKey) {
    return config().toBuilder()
      .features(Collections.singletonMap(featureKey, true))
      .build();
  }

  public static @NonNull Config configWithFeaturesEnabled(final @NonNull Map<String, Boolean> features) {
    return config().toBuilder()
      .features(features)
      .build();
  }
}

File: app/src/main/java/com/kickstarter/mock/factories/ConfigFactory.kt
package com.kickstarter.mock.factories

import com.kickstarter.libs.Config
import com.kickstarter.libs.Config.LaunchedCountry

object ConfigFactory {
    @JvmStatic
    fun config(): Config {
        val US = LaunchedCountry.builder()
            .name("US")
            .currencyCode("USD")
            .currencySymbol("$")
            .trailingCode(true)
            .build()
        val GB = LaunchedCountry.builder()
            .name("GB")
            .currencyCode("GBP")
            .currencySymbol("£")
            .trailingCode(false)
            .build()
        val CA = LaunchedCountry.builder()
            .name("CA")
            .currencyCode("CAD")
            .currencySymbol("$")
            .trailingCode(true)
            .build()
        val JP = LaunchedCountry.builder()
            .name("JP")
            .currencyCode("JPY")
            .currencySymbol("¥")
            .trailingCode(false)
            .build()
        return Config.builder()
            .countryCode("US")
            .features(mutableMapOf())
            .launchedCountries(listOf(US, GB, CA, JP))
            .build()
    }

    fun configForUSUser(): Config {
        return config()
            .toBuilder()
            .countryCode("US")
            .build()
    }

    fun configForCA(): Config {
        return config().toBuilder()
            .countryCode("CA")
            .build()
    }

    fun configForITUser(): Config {
        return config().toBuilder()
            .countryCode("IT")
            .build()
    }

    fun configWithExperiment(experiment: String, variant: String): Config {
        return config().toBuilder()
            .abExperiments(
                mutableMapOf<String, String>().apply {
                    this[experiment] = variant
                }
            )
            .build()
    }

    fun configWithExperiments(abExperiments: Map<String, String>): Config {
        return config().toBuilder()
            .abExperiments(abExperiments)
            .build()
    }

    fun configWithFeatureEnabled(featureKey: String): Config {
        return config().toBuilder()
            .features(
                mutableMapOf<String, Boolean>().apply {
                    this[featureKey] = true
                }
            )
            .build()
    }

    fun configWithFeatureDisabled(featureKey: String): Config {
        return config().toBuilder()
            .features(
                mutableMapOf<String, Boolean>().apply {
                    this[featureKey] = false
                }
            )
            .build()
    }

    fun configWithFeaturesEnabled(features: Map<String, Boolean>): Config {
        return config().toBuilder()
            .features(features.toMutableMap())
            .build()
    }
}
